torch/nn/modules/module.py

```diff
@@ -855,7 +873,14 @@ def _slow_forward(self, *input, **kwargs):
                 tracing_state.pop_scope()
         return result
 
+    has_hooks = None
+
     def _call_impl(self, *input, **kwargs):
+        if global_hooks_updated.value or self._module_hooks_updated.value:
+            self.has_hooks = max(len(self._backward_hooks), len(_global_backward_hooks), len(_global_forward_hooks), len(self._forward_hooks), len(_global_forward_pre_hooks), len(self._forward_pre_hooks), len(_global_forward_pre_hooks)) == 0
+
+        if not self.has_hooks and not torch._C._get_tracing_state():
+            return self.forward(*input, **kwargs)
```

Comment:
While I see why this happens here, I fear this is fairly brittle code as it actually duplicate (in a non-trivial way) logic from the function below. meaning that some updates to the function below will need to be replicated here.